### PR TITLE
Clean up thread_management.h

### PR DESCRIPTION
**Pull request by masterleinad**
This PR refactors the various overloads for `new_thread` and `new_task` in two variadic template versions. 
Further, it turned out out that the `internal` structs `fun_ptr_helper` and `fun_encapsulator` were unused
and can hence be deleted.

### CODE REVIEW

**Comment by bangerth** [comment]
Nice job deleting all of this code. Have you checked that everything still works if you disable threading?

There is currently one test that fails (`fe_system_generalized_support_points`) but that is a failure due to a different patch and unrelated to the current one.

File: include/deal.II/base/thread_management.h [hunk 1]
```diff
@@ -1823,2344 +1823,701 @@ namespace Threads
...
-  }
-
-#ifndef DEAL_II_CONST_MEMBER_DEDUCTION_BUG
-  /**
-   * Overload of the new_task function for const member functions with 1
-   * argument.
-   *
-   * @ingroup threads
-   */
-  template <typename RT, typename C,
```
Comment (bangerth) [inline include/deal.II/base/thread_management.h]: nice!

**Comment by bangerth** [comment]
See also #2300 and #4249.

**Comment by bangerth** [comment]
Hm, now I worry that I was too quick with merging. I would appreciate if you could try running everything with threads disabled and see whether things still compile. :-(

**Comment by masterleinad** [comment]
Yes, I can compile with `DEAL_II_:WITH_THREADS=OFF`.

**Comment by bangerth** [comment]
Thanks for double checking! Amazing how we had these hundreds of lines of classes that are no longer used just sitting around there.

**Pull request merged**